src/System.Windows.Forms/src/System/Windows/Forms/Control.cs

```
@@ -10166,7 +10167,7 @@ internal void SetAcceptDrops(bool accept)
                         Debug.WriteLineIf(CompModSwitches.DragDrop.TraceInfo, "   ret:" + n.ToString(CultureInfo.CurrentCulture));
                         if (n != HRESULT.S_OK && n != HRESULT.DRAGDROP_E_ALREADYREGISTERED)
                         {
-                            throw Marshal.GetExceptionForHR((int)n);
+                            throw Marshal.GetExceptionForHR((int)n)!;
```

We don't know this. It could be `null` ex: S_FALSE.  We cannot use   `ThrowExceptionForHR` either.  We may need to customize in such cases here. @Tanya-Solyanik  any suggestions?

---

Special-case S_FALSE.

---

@gpetrou  can you see if this is the only constant that could thow null here? or any positive value end up throwing null?

---

I assume you are referring to https://github.com/dotnet/runtime/blob/215b39abf947da7a40b0cb137eab4bceb24ad3e3/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L643?
In previous PRs, we have used the null forgiving operator with `GetExceptionForHR`. Do you want to throw a different exception if null is returned?

---

`RegisterDragDrop` does not seem to return S_FALSE though as indicated in https://docs.microsoft.com/en-us/windows/win32/api/ole2/nf-ole2-registerdragdrop#return-value

---

> I assume you are referring to https://github.com/dotnet/runtime/blob/215b39abf947da7a40b0cb137eab4bceb24ad3e3/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L643? In previous PRs, we have used the null forgiving operator with `GetExceptionForHR`. Do you want to throw a different exception if null is returned?

yes i was referring to it.  In such case, we may want to just return error code instead of mapping. Did you check any of those error codes mentioned in the document for `RegisterDragDrop` has positive values?

---

This is an existing code, so I'm not sure we should start rewriting this....
We also have this extension:
https://github.com/dotnet/winforms/blob/97777ed2310f23a0f6662540e9e393dba922de57/src/System.Windows.Forms.Primitives/src/Interop/Interop.HRESULT.cs#L94-L100

---

We have the following values, so I assume that we can just keep the current changes.
| Name | Value |
| ---: | :--- |
| DRAGDROP_E_INVALIDHWND | 0x80040102 |
| DRAGDROP_E_ALREADYREGISTERED | 0x80040101 |
| E_OUTOFMEMORY | 0x8007000E |

---

I wish we had a better solution but do not want to rewrite this part given this has no impact at runtime behavior, i am fine leaving it as-is.